From 30d0b8ab05cf8f940db3eacd359333278a8aabee Mon Sep 17 00:00:00 2001
From: Shuguang Gong <shuguang.gong@intel.com>
Date: Wed, 3 Jun 2015 15:45:24 +0800
Subject: [PATCH 412/462] ov2685: enable AWB lock

AWB lock needs to be enabled since sensor supports it

Change-Id: I52073ed29b30176c98e2f9ea2621957ee3128219
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-37223
Signed-off-by: Shuguang Gong <shuguang.gong@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/58678
Validation-Android: Wei Tao <wei.tao@intel.com>
Approver: Tian Shu Qiu <tian.shu.qiu@intel.com>
Build: IMIN (SC)
Integrator: WeiX Meng <weix.meng@intel.com>
Build: WeiX Meng <weix.meng@intel.com>
Reviewed-by: WeiX Meng <weix.meng@intel.com>
Maintainer: WeiX Meng <weix.meng@intel.com>
---
 drivers/external_drivers/camera/drivers/media/i2c/ov2685.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov2685.c b/drivers/external_drivers/camera/drivers/media/i2c/ov2685.c
index bc56309..7e0ad0e 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov2685.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov2685.c
@@ -359,6 +359,10 @@ static int __ov2685_s_wb(struct v4l2_subdev *sd, int wb)
 	struct ov2685_device *dev = to_ov2685_sensor(sd);
 	int ret;
 
+	/* Not to set awb when awb is locked */
+	if (dev->ae_lock & V4L2_LOCK_WHITE_BALANCE)
+		return 0;
+
 	switch (wb) {
 	case V4L2_WHITE_BALANCE_MANUAL:
 		ret = ov2685_write_reg_array(client, ov2685_AWB_manual);
@@ -712,7 +716,16 @@ static int ov2685_g_ae_lock(struct v4l2_subdev *sd, s32 *value)
 
 static int ov2685_s_ae_lock(struct v4l2_subdev *sd, int value)
 {
+	struct i2c_client *client = v4l2_get_subdevdata(sd);
 	struct ov2685_device *dev = to_ov2685_sensor(sd);
+	int ret;
+
+	/* only awb lock is supported by ov2685 */
+	if (value & V4L2_LOCK_WHITE_BALANCE) {
+		ret = ov2685_write_reg_array(client, ov2685_AWB_manual);
+		if (ret)
+			return ret;
+	}
 
 	dev->ae_lock = value;
 	return 0;
-- 
1.9.1

